homeassistant/core.py

```diff
@@ -748,7 +763,7 @@ def async_fire(
                     continue
             self._hass.async_add_hass_job(job, event)
 
-    def listen(self, event_type: str, listener: Callable) -> CALLBACK_TYPE:
+    def listen(self, event_type: str, listener: _ListenerType) -> CALLBACK_TYPE:
```

Comment:
Will the alias be expanded to its parts in eg vscode type hinting? Aliases makes it harder to see what the actual types are, when reading the code, so we should be careful with them.

Reply:
Yeah, I agree, I'm not a fan of aliases in these places.

Reply:
I understand your concerns, it's a tradeoff in this case. One I believe to be worth it though.

> Will the alias be expanded to its parts in eg vscode type hinting?

No, at least not initially. There have been multiple discussions about that topic in the Pylance repo but the conclusion has been that the signature is more readable if they aren't expanded. An example, note that `HassJob` will only become a Generic in #63245

<img width="618" alt="Screen Shot 2022-01-03 at 12 43 41" src="https://user-images.githubusercontent.com/30130371/147926915-ff433cc1-d6ae-49e2-8ca9-e5e15b374e61.png">

<img width="627" alt="Screen Shot 2022-01-03 at 12 45 24" src="https://user-images.githubusercontent.com/30130371/147926890-0acf285d-102c-4a5b-b368-04fe82e62a8d.png">

In VS Code, you can always use `Go to definition` and then however over the alias to the see full type.

> Aliases makes it harder to see what the actual types are, when reading the code, so we should be careful with them.

True, but they make the signature much more readable. In the example above, it's way easier to see that only `event_type` and `listener` are required attributes whereas `event_filter` can be `None`.

Reply:
> In VS Code, you can always use Go to definition and then however over the alias to the see full type.

The fact that that is needed is not good and the concern raised. It means that the code isn't readable.

Reply:
> > In VS Code, you can always use Go to definition and then however over the alias to the see full type.
> 
> The fact that that is needed is not good and the concern raised. It means that the code isn't readable.

It's a tradeoff like I said. IMO the second screenshot isn't any more helpful. I would also have to refer back to the original implementation to fully understand the signature but you might disagree.

Some last arguments in favor. It's much easier to keep a complex type in sync if it's only defined once. Especially with `Callable` and `Generic` types, you get to a point where it's necessary to deal with a lot of nested brackets and unions. After all complex types are exactly what TypeAliases are for. I'm also not taking about replacing every type definition. As far as I remember, these three for `EventBus` are the only once I've added.

If the consensus still is that we shouldn't use them here, I'll change it. Although I hope that I could have convince you that they are actually helpful in this case.

Reply:
I think `_ListenerType` is not necessary. It is simple enough and we didn't do it for the similar `ServiceCall` in #63241
Similarly, I don't think `_EventFilterType` adds much value.

Maybe only implement `_FilterableJobType`?

Reply:
> I think `_ListenerType` is not necessary. It is simple enough and we didn't do it for the similar `ServiceCall` in #63241

Personally, I would recommend it for `ServiceCall` too, but haven't done so yet as it was `Callable[[ServiceCall], Any]` previously. With the last change to the return type, it could make sense now.

> Similarly, I don't think `_EventFilterType` adds much value. Maybe only implement `_FilterableJobType`?

On its own maybe not, but it helps a lot when trying to understand `_FilterableJobType`. In the end this type will be fairly complex. Without the alias, it will just get even more difficult to read.
```py
_FilterableJobType = Tuple[
    HassJob[Union[None, Awaitable[None]]], Union[_EventFilterType, None]
]
```
```py
_FilterableJobType = Tuple[
    HassJob[Union[None, Awaitable[None]]], Union[Callable[[Event], bool], None]
]
```

Reply:
Our python requirements are new enough that we could now make `filterable_job` a `dataclass` or similar.

Reply:
> Our python requirements are new enough that we could now make `filterable_job` a `dataclass` or `namedtuple` or ....

Good point! I've added a new `NamedTuple` class `_FilterableJob` for it. Although a NamedTuple is basically an Alias itself, it definitely helps with readability.

My previous points still stand though. The aliases for `_ListenerType` and `_EventFilterType` IMO improve readability of the method signature and help with getting the types right in various locations by only defining it once.

Reply:
`_EventFilterType` is so simple now: `Callable[[Event], bool]`

Why even bother aliasing that?

Reply:
> Why even bother aliasing that? I'm sorry... This is not my style or something I can get used too I think.

I guess we have a difference of opinion on that. Anyway, I removed `_EventFilterType` and `_ListenerType`. Would it be fine with you if the `_FilterableJob` NamedTuple stays? It's only used in `core.py`.

Reply:
> Would it be ok for you if the _FilterableJob NamedTuple stays?

👍 

I think the end result looks alright 👍